lib/collection/src/shards/local_shard/shard_ops.rs

```diff
@@ -261,23 +262,82 @@ impl ShardOperation for LocalShard {
         timeout: Option<Duration>,
         hw_measurement_acc: HwMeasurementAcc,
     ) -> CollectionResult<Vec<ShardQueryResponse>> {
-        let planned_query = PlannedQuery::try_from(requests.as_ref().to_owned())?;
+        // Chunk requests for parallelism in certain scenarios
+        //
+        // Deeper down, each segment gets its own dedicated search thread. If this shard has just
+        // one segment, all requests will be executed on a single thread.
+        //
+        // To prevent this from being a bottleneck if we have a lot of requests, we can chunk the
+        // requests into multiple searches to allow more parallelism.
+        //
+        // We don't create smaller chunks than MIN_BATCH_SIZE to ensure our 'filter reuse
+        // optimization' can be properly used.
+        // See: <https://github.com/qdrant/qdrant/pull/813>
+        // See: <https://github.com/qdrant/qdrant/pull/6326>
+        const MIN_CHUNK_SIZE: usize = 10;
+        let chunk_size = {
+            let count = requests.len();
+            if count <= MIN_CHUNK_SIZE {
+                // If number of requests is less than minimum batch size, create one chunk
+                count
+            } else {
+                let segment_count = max(self.segments.read().len().saturating_sub(1), 1);
+                if count <= segment_count {
+                    // If number of requests is less than segment count, create one chunk
+                    // Each request is also already parallelized over each segment
+                    count
+                } else {
+                    // Determine CPU count per segment, create up to that amount of chunks to parallelize
+                    // Each request is also already parallelized over each segment
+                    let cpus_per_segment = max(common::cpu::get_num_cpus() / segment_count, 1);
```

Comment:
Number of CPUs dedicated for search is explicitly configured https://github.com/qdrant/qdrant/blob/7f3a2c298135cae23da4a596379f2db74dd24cbc/src/main.rs#L270-L269